File: dzo_composer.pl
#!/usr/bin/perl -wTU
#****************************************************************************************
use strict;
use warnings;
use CGI qw{ :standard };
use CGI::Carp qw ( fatalsToBrowser );
use File::Basename;
use File::Path qw(make_path);
use Image::Magick;
use POSIX qw(ceil);
use Time::HiRes qw(gettimeofday tv_interval);
#****************************************************************************************
my $query = new CGI;
my $safe_filename_characters = "a-zA-Z0-9_.-";
my $upload_dir = $query->param("upload_dir"); #Using precreated dirs, tiles and info will be created under this dir.
	$upload_dir =~  m/^([a-zA-Z0-9\._\/\-]+)$/ or Error(0,"Tainted $upload_dir","$!");
	$upload_dir = $1;
my $filename = $query->param("file");
	$filename =~ s/(?:\\|\/)([^\\\/]+)$/$1/g;
my ( $name, $path, $extension ) = fileparse ( $filename, '..*' );
	$filename = $name . $extension;
my $upload_filehandle = $query->upload("file");
my $overlap = $query->param("overlap");
my $tilesize = $query->param("tilesize");
my $filesize_real = $query->param("file_size"); #The real file size, not the form data one
my $sessid = $query->param("sessid");
	$sessid =~  m/^([a-zA-Z0-9\._]+)$/ or Error(0,"Tainted $sessid","$!");
my $pingfile = "$1.log";
my $debug_info = $query->param("debug_info"); #Pass additional debug info to the referrer.
my $log_info = $query->param("log_info"); #write additional log info.
my $t0 = [gettimeofday]; #Vars used to stat timelapses
my $elapsed = tv_interval ($t0);
#****************************************************************************************/^(\w+)$/
MAIN:
{
print "Content-type: text/html\n\n";
if ($log_info eq "false")
{
	$log_info = "true";
	printInfo("MAIN");
	$log_info = "false";
}
else{printInfo("MAIN");}
&Proceso;
&Fin;
exit 0;
}
#****************************************************************************************
sub Proceso
{
printInfo("Proceso");
open ( UPLOADFILE, ">$upload_dir/$filename" ) or  Error(2,"Tainted $upload_dir","$!");;
binmode UPLOADFILE;
while ( <$upload_filehandle> )
{
	print UPLOADFILE;
}
close UPLOADFILE;
printInfo("File uploaded and written - Proceso");
DZI_creation();
html_examples();
return;
}
sub html_examples
{
printInfo("Parsing example - Proceso");
printInfo("Example parsed - Proceso");
return;
}
sub DZI_creation
{
	printInfo("generate - DZI_creation started - Proceso");
	$t0 = [gettimeofday]; #start time profiler
	# Create the tiles
	#Parameters
	#Image, path ,.dzi name, output format, tilesize, overlap, width, height, zipped output, reason, progress(fine/coarse)
	my $image = new Image::Magick;
	$image->Read("$upload_dir/$filename");
	my $tempImage = $image->Clone();
	my $croppedImage = $tempImage->Clone();
	my ($width, $height) = $image->Get('width', 'height');#Get dimensions of the image.
	my $maxDimension = ($width, $height)[$width < $height]; #Get the larger side
	#Create DZI and Json info
	my $dziStr = "<?xml version='1.0' encoding='UTF-8'?><Image xmlns='http://schemas.microsoft.com/deepzoom/2008' Format='jpg' Overlap='$overlap' TileSize='$tilesize' ><Size Height='$height' Width='$width' /></Image>";
	my $jsonStr = "{Image: {xmlns: 'http://schemas.microsoft.com/deepzoom/2008', Format: 'jpg', Overlap: $overlap, TileSize: $tilesize, Size: { Width: $width , Height: $height } }}";
	printFile("$upload_dir/$filename.dzi",">",$dziStr);
	printFile("$upload_dir/$filename.json",">",$jsonStr);
	#Calculate resize steps. Bassically calculate how many times we need to elevate 2 to get the goal size.
	my $stepsmax = int((log($maxDimension)/log(2)) + 1);

	my $tempX = $width;
	my $tempY = $height;
	my $steps = $stepsmax;
	my $stepsX = 0;
	my $stepsY = 0;
	my $j = 0;
	my $k = 0;
	my $X1 = 0;
	my $Y1 = 0;
	my $stro = "";
	for (my $i = 0; $i <= $stepsmax; $i++)
	{
		#Create a clone to use as a temp image.
		$tempImage = $image->Clone();
		#First step no resize.
		if ($steps == $stepsmax)
		{
			printInfo("First Step:  $steps  \n");
			($width, $height) = $image->Get('width', 'height');
			$stepsX = int(($width/$tilesize));
			$stepsY = int(($height/$tilesize));
			printInfo("First dimensions.Tilesize: $tilesize TempX: $tempX X: $width Y: $height\n");
		}
		else #Resize the temp image
		{
			printInfo("Next Steps: $steps\n");
			$tempX = int(ceil($tempX / 2));
			$tempY = int(ceil($tempY / 2));
			$tempImage->Scale(width=>$tempX, height=>$tempY);
			($width, $height) = $tempImage->Get('width', 'height');
			$stepsX = int(($width/$tilesize));
			$stepsY = int(($height/$tilesize));
			printInfo("$sessid $filename Next dimensions. TempX: $tempX X: $width Y: $height\n");
		}
		#while cropped width > tilesize
		for ($j = 0; $j <= $stepsX; $j++)
		{
			$X1=($tilesize*$j)-($overlap*$j);
			for ($k = 0; $k <= $stepsY; $k++)
			{
				if (-e "$upload_dir/$filename\_files/$steps/$j\_$k.jpg")
				{
					print "$upload_dir/$filename\_files/$steps/$j\_$k.jpg already Exists";
					printFile("$pingfile", ">", "$sessid:$i:$stepsmax\n");
				}
				else
				{
			#Create a clone of the resized clone to use as a temp image for every crop done.
					$Y1=($tilesize*$k)-($overlap*$k);
					$croppedImage = $tempImage->Clone();
					$croppedImage->Crop(width=>$tilesize,height=>$tilesize,x=>$X1,y=>$Y1);
					($width, $height) = $croppedImage->Get('width', 'height');
					make_path("$upload_dir/$filename\_files/$steps");
					$croppedImage->Write(filename=>"$upload_dir/$filename\_files/$steps/$j\_$k.jpg", compression => 'JPEG', quality => '85');
					$stro = "File: $filename step: $i $steps $stepsmax stepX: $j $stepsX StepY: $k $stepsY X1: $X1 Y1: $Y1 cropped width: $width cropped height: $height\n";
					printInfo("$sessid $filename $stro");
					printFile("$pingfile", ">", "$sessid:$i:$stepsmax\n");
				}
			}
		}
	$steps--;
	}
	$elapsed = tv_interval ($t0); #Finish profiling
	my @dateandtime = Datetime();
	printFile("upload.log.txt",">>","@dateandtime DZI_creation ended - Proceso. Time to complete: $elapsed.\n");
return;
}
sub Untantizeme
{
my ($fichero) = @_;
	if ($fichero =~ /^([-\@\w.]+)$/)
	{
		$fichero = $1; 			# $data now untainted
	} else
	{
		 Error(0,"Tainted $fichero","$!"); 	# log this somewhere
	}	
return $fichero;
}
#****************************************************************************************
sub Fin
{
my ($Mensaje, $Parametro) = @_;
if ($log_info eq "false")
{
	$log_info = "true";
	printInfo("Fin");
	$log_info = "false";
}
else {printInfo("Fin");}
print "<HTML><HEAD><TITLE></TITLE>"; 
print print "</HEAD><BODY BGCOLOR='#FFFFFF' TEXT='#000000' LINK='#ff80ff' VLINK='#ff80ff' ALINK='#ff80ff'>Proceso finalizado</BODY></HTML>"; 
exit 0;
}
#*******************************************************************************

sub Error
{
my ($criticidad, $Mensaje, $Parametro) = @_;
my @dateandtime = Datetime();
	printInfo("Error - $Mensaje, $Parametro");
	print "Content-type: text/html\n\n";
	print "<HTML><HEAD><TITLE>$Mensaje</TITLE>";
	print "</HEAD><BODY BGCOLOR='#FFFFFF' TEXT='#000000' LINK='#ff80ff' VLINK='#ff80ff' ALINK='#ff80ff'>$dateandtime[0] - $dateandtime[1]<br />Ha habido un error, $Mensaje: $Parametro.</BODY></HTML>"; 
	if ($criticidad > 0){exit 0;}
	else {return;}
}

sub Datetime
{
my ($sec,$min,$hour,$mday,$mon,$year,$wday,$yday,$isdst) = localtime(time);
my $Ano  = $year + 1900;
my $Mes = $mon + 1;
my $Dia = $mday;
my $Decimal;
for ($Decimal = 0 ; $Decimal < 10 ;$Decimal++)
	{
	if ($mday eq "$Decimal") {$Dia = join("","0",$mday);}
	if ($Mes eq "$Decimal") {$Mes = join("","0",$Mes);}
	}
my $Fecha = join ("\/",$Ano,$Mes,$Dia);
my $Hora = join (":",$hour,$min,$sec);
if ($Fecha eq ""){$Fecha = "50000000";}
my @dateandtime = ($Fecha,$Hora);
return @dateandtime;
}
sub printFile
{
my ($fileName, $mode, $fileContents) = @_;
open (FILE,"$mode$fileName") || Error(1,"can't open $filename","$!");
	print FILE $fileContents;
close (FILE);
return;
}
sub printInfo
{
my ($Mensaje) = @_;
my @dateandtime = Datetime();
if ($debug_info eq "true")
	{
	print "$dateandtime[0] - $dateandtime[1] - $ENV{'HTTP_REFERER'} - $filename - $filesize_real - $sessid - $Mensaje\n<BR/>";
	}
if ($log_info eq "true")
	{
	open (LOG,'>>upload.log.txt') || Error(1,"can't open >>upload.log.txt","$!");
		print LOG "$dateandtime[0] - $dateandtime[1] - $ENV{'HTTP_REFERER'} - $filename - $filesize_real - $sessid - $Mensaje\n";
	close (LOG);
	}
return;
}

File: dzo_composer.php
<?php
//myDeepZoom very nice Composer of the happy ardent OpenSeadragon (DZO) v0.1
//Part of the OpenSeadragon DZI online composer utils.
//Needs imagick.so installed and active in the PHP settings.
//Alfonso Abelenda Escudero, chafalladas.com. (c) 2016 license BSD-type 2.
//In fact is my first PHP script ever, needs a lot of recode and cleansing.
//It lacks functions and a proper structure since I was learning how to use PHP.
//But it does the job for the good local people.
global $debug_info, $log_info, $sessid;
function prof_flag($str)
{
    global $prof_timing, $prof_names, $sessid;
    $prof_timing[] = microtime(true);
    $prof_names[] = $str;
}
function prof_print()
{
    global $prof_timing, $prof_names, $sessid;
    $size = count($prof_timing);
    for($i=0;$i<$size - 1; $i++)
    {
		FilePrint("log_dzo.txt", "",date(DATE_RFC2822) . " " . $sessid . " " . $target_file .  $prof_names[$i] . ": ","a+", true);
		FilePrint("log_dzo.txt", "", sprintf("%f\n", $prof_timing[$i+1]-$prof_timing[$i]),"a+", true);
    }
}
prof_flag("Error Handling");
function myErrorHandler($errno, $errstr, $errfile, $errline)
{
global $sessid;
//Take care of known warning so they don't get logged
$myWarns = array("Undefined variable: tileSize","File exists");
foreach ($myWarns as $warn){$pos = strpos($errstr, $warn);$warn = $warn + $pos;}
if($warn !== false)
{
	return;
}
else
{
    switch ($errno)
	{
    case E_USER_ERROR:
        echo "<b>My ERROR</b> [$errno] $errstr<br />\n";
        echo "  Fatal error on line $errline in file $errfile";
        echo ", PHP " . PHP_VERSION . " (" . PHP_OS . ")<br />\n";
        echo "Aborting...<br />\n";
		FilePrint("log_dzo.txt", "",date(DATE_RFC2822) . " " . $sessid . " " . $target_file .  "My ERROR</b> [$errno] $errstr at $errline in $errfile\n","a+", true);
        exit(1);
        break;
    case E_USER_WARNING:
        echo "My WARNING</b> [$errno] $errstr<br />\n";
		FilePrint("log_dzo.txt", "",date(DATE_RFC2822) . " " . $sessid . " " . $target_file .  "My WARNING</b> [$errno] $errstr at $errline in $errfile\n","a+", true);
        break;
    case E_WARNING:
        echo "<b>My WARNING</b> [$errno] $errstr<br />\n";
		FilePrint("log_dzo.txt", "",date(DATE_RFC2822) . " " . $sessid . " " . $target_file .  "My WARNING</b> [$errno] $errstr at $errline in $errfile\n","a+", true);
        break;
    case E_USER_NOTICE:
        echo "<b>My NOTICE</b> [$errno] $errstr<br />\n";
		FilePrint("log_dzo.txt", "",date(DATE_RFC2822) . " " . $sessid . " " . $target_file .  "My NOTICE</b> [$errno] $errstr at $errline in $errfile\n","a+", true);
        break;
    default:
        echo "Unknown error type: [$errno] $errstr<br />\n";
		FilePrint("log_dzo.txt", "",date(DATE_RFC2822) . " " . $sessid . " " . $target_file .  "Unknown error type: [$errno] $errstr at $errline in $errfile\n","a+", true);
        break;
    }
    /* Don't execute PHP internal error handler */
    return true;
}
}
$old_error_handler = set_error_handler("myErrorHandler",E_ALL | E_WARNING);
$target_dir = $_POST['upload_dir']; #Using precreated dirs, tiles and info will be created under this dir.
$sessid = $_POST['sessid'];
$debug_info = $_POST['debug_info'];
$log_info = $_POST['log_info'];
$overlap = $_POST['overlap'];
$tilesize =  $_POST['tilesize'];
$target_file = $target_dir . "/" . basename($_FILES["file"]["name"]);
$uploadOk = 1;
$imageFileType = pathinfo($target_file,PATHINFO_EXTENSION);
//my $log_info = $query->param("log_info"); #Pass log info to the referrer.
prof_flag("Start");
FilePrint($sessid, ".txt", $sessid . ":0:12\n","w", true);
//print_r(get_declared_classes()); //Good for testing the classes loaded.

// Check if image file is a actual image or fake image
if(isset($_POST["submit"])) {
    $check = getimagesize($_FILES["file"]["tmp_name"]);
    if($check !== false) {
		FilePrint("log_dzo.txt", "",date(DATE_RFC2822) . " " . $sessid . " " . $target_file .  "File is an image - " . $check["mime"] . ".\n","a+", true);
        $uploadOk = 1;
    } else {
		FilePrint("log_dzo.txt", "", date(DATE_RFC2822) . " " . $sessid . " " . $target_file . "File is not an image.\n","a+", true);		
        $uploadOk = 0;
    }
}

// Check if $uploadOk is set to 0 by an error
if ($uploadOk == 0) {
	FilePrint("log_dzo.txt", "", date(DATE_RFC2822) . " " . $sessid . " " . $target_file . "Sorry, there was an error uploading your file.\n","a+", true);
	exit(1);
	// if everything is ok, try to upload file
} else {
    if (move_uploaded_file($_FILES["file"]["tmp_name"], $target_file)) {
		FilePrint("log_dzo.txt", "", date(DATE_RFC2822) . " " . $sessid . " " . $target_file . "The file ". basename( $_FILES["file"]["name"]). " has been uploaded.\n","a+", true);
    } else {
		FilePrint("log_dzo.txt", "", date(DATE_RFC2822) . " " . $sessid . " " . $target_file . "Sorry, there was an error uploading your file.\n","a+", true);
		exit(1);
    }
}
// Create the tiles
//Parameters
//Image, path ,.dzi name, output format, tilesize, overlap, width, height, zipped output, reason, progress(fine/coarse)
prof_flag("Tiling");
try
{
        /*** a valid image file ***/
        $image = $target_file;

        /*** a new imagick object ***/
		$myimage = new Imagick($image);
}
catch(Exception $e)
{
        echo $e->getMessage();
		echo '\n';
}
//Get dimensions of the image.
$d = $myimage->getImageGeometry();
$maxDimension = max($d['height'],$d['width']);
//Create DZI and Json info
$dziStr = "<?xml version='1.0' encoding='UTF-8'?><Image xmlns='http://schemas.microsoft.com/deepzoom/2008' Format='jpg' Overlap='" . $overlap . "' TileSize='" . $tilesize . "' ><Size Height='" . $d['height'] . "' Width='" . $d['width'] . "' /></Image>";
$jsonStr = "{Image: {xmlns: 'http://schemas.microsoft.com/deepzoom/2008', Format: 'jpg', Overlap:" . $overlap . ", TileSize: " . $tileSize . ", Size: { Width: " . $d['width'] . " , Height: " . $d['height'] . " } }}";
FilePrint($target_file,".dzi",$dziStr,"w", true);
FilePrint($target_file,".json",$jsonStr,"w", true);
//Append tilesets to the json file of processed images.
$jsonStr = ",\n{myDirTilesource:'" . $target_dir ."',mytileName:'" . $target_file . ".dzi',tx:1,ty:0, tHeight: " . $d['height'] . ", tWidth: " . $d['width'] . "}]}";
//Calculate resize steps. Bassically calculate how many times we need to elevate 2 to get the goal size.
$stepsmax = (int)(log($maxDimension,2)) + 1;
//We choose to draw the less steps.
//While smaller side > 1
//$steps = $stepsmax;
$tempX = $d['width'];
$tempY = $d['height'];
$steps = $stepsmax;
for ($i = 0; $i <= $stepsmax; $i++)
{
	//Create a clone to use as a temp image.
	$tempImage = clone $myimage;
	//First step no resize.
	if ($steps == $stepsmax)
	{
		FilePrint("log_dzo.txt","",date(DATE_RFC2822) . " " . $sessid . " " . $target_file . " First Step: " . $steps . "\n","a+", true);
		$d = $tempImage->getImageGeometry();
		$stepsX = (int)($d['width']/$tilesize);
		$stepsY = (int)($d['height']/$tilesize);
		FilePrint("log_dzo.txt","",date(DATE_RFC2822) . " " . $sessid . " " . $target_file . "First dimensions. TempX: ". $tempX ." X: " . $d['width'] . " Y: " . $d['height'] . "\n","a+", false);
	}
	else //Resize the temp image
	{
		FilePrint("log_dzo.txt","",date(DATE_RFC2822) . " " . $sessid . " " . $target_file . "Next Steps: " . $steps . "\n", "a+", false);
		$tempX = (int)ceil($tempX / 2);
		$tempY = (int)ceil($tempY / 2);
		$tempImage->scaleImage($tempX, $tempY, true);
		$d = $tempImage->getImageGeometry();
		$stepsX = (int)($d['width']/$tilesize);
		$stepsY = (int)($d['height']/$tilesize);
		FilePrint("log_dzo.txt","",date(DATE_RFC2822) . " " . $sessid . " " . $target_file . "Next dimensions. TempX: ". $tempX ." X: " . $d['width'] . " Y: " . $d['height'] . "\n","a+", false);
	}
	//while cropped width > tilesize
	for ($j = 0; $j <= $stepsX; $j++)
	{
		$X1=($tilesize*$j)-($overlap*$j);
		for ($k = 0; $k <= $stepsY; $k++)
		{
	//Create a clone of the resized clone to use as a temp image for every crop done.
			$Y1=($tilesize*$k)-($overlap*$k);
			$croppedImage = clone $tempImage;
			$croppedImage->cropImage($tilesize,$tilesize, $X1, $Y1);
			$d = $croppedImage->getImageGeometry();
			mkdir($target_file . "_files/", 0755);
			mkdir($target_file . "_files/" . $steps, 0755);
			//Write the cropped file in the correspondent dir.
			$croppedImage->writeImage($target_file . "_files/" . $steps . "/" . $j . "_". $k . ".jpg");
			$stro = "File: " . $target_file . " step: " . $steps . " stepX: " . $j . " StepY: ". $k ." X1: " . $X1 . " Y1: " . $Y1 . " cropped width: " . $d['width'] . " cropped height: " . $d['height'] . "\n";
			FilePrint("log_dzo.txt","",date(DATE_RFC2822) . " " . $sessid . " " . $target_file . $stro,"a+", false);
			FilePrint($sessid, ".txt", $sessid . ":" . $i . ":" . $stepsmax . "\n", "w", true);
		}
	}
$steps--;
}
prof_flag("Done");
prof_print();
//Do a happy day regard to the user.
function FilePrint($file,$ext,$string,$mode,$mandatory)
{
global $debug_info, $log_info;
	if ($debug_info == "true"){echo $string;}
	if ($log_info == "true" || $mandatory == true)
	{
		$fp = fopen($file . $ext, $mode);
		fwrite($fp, $string);
		fclose($fp);
	}	
}
?>